### ISSUE TITLE
Option for permanent preselect

### ISSUE THREAD

**Issue by oblitum** [neoclide/coc.nvim#3172]
**Is your feature request related to a problem? Please describe.**

Current preselect feature depends on server to send preselected item, but it's the case that few servers do it, gopls is one exception.

Initially the preselect feature in coc.nvim was default enabled, but it moved to default disabled out of confusion it generated (possibly a lot due to the server reliance problem).

**Describe the solution you'd like**

I wish a continuous preselect of first completion item, not mattering whether server provides preselect or not. And maybe additional variation to provide continuous preselect in case the server doesn't report preselect, and in case it does, pick from server.

**Describe alternatives you've considered**

**Additional context**

nvim-compe provides similar option: https://github.com/hrsh7th/nvim-compe/blob/faf81b898659fcf67139251868500c6e6d62a088/doc/compe.txt#L247-L255

**Comment by kevinhwang91** [comment]
```json
	"suggest.enablePreselect": true,
	"suggest.noselect": false,
```
will work, but I expect the preselected item always become the first and use the new kind label.

**Comment by chemzqm** [comment]
Disable `"suggest.noselect"` does what you want, the reason preselect is disabled is that many users don't like complete item automatic selected on shown, which possible cause unnecessary redraw and completion resolve.

**Comment by chemzqm** [comment]
> ```json
> 	"suggest.enablePreselect": true,
> 	"suggest.noselect": false,
> ```
> 
> will work, but I expect the preselected item to always become the first and use the new shortcut or kind label.

The position of preselected item could be not the first, the same as VSCode.
We could add additional character to the end of `abbr`.

**Comment by oblitum** [comment]
Thanks for the info. Didn't know setting these two that way would enable this.

> many users don't like complete item automatic selected on shown, which possible cause unnecessary redraw and completion resolve

TBH, I just wanted to have the option available, I possibly will end up not using it due to that.

**Comment by oblitum** [comment]
I'm trying it, and it's working like the second variation I mentioned indeed ("And maybe additional variation to provide continuous preselect in case the server doesn't report preselect, and in case it does, pick from server"), though I think would be nice have just first item preselect option.

**Comment by kevinhwang91** [comment]
@chemzqm If you don't want to make this behavior different from VSC, publish a new interface for completion results to users is a good choice. User can wrap their completion strategy.

**Comment by chemzqm** [comment]
@kevinhwang91 what interface?

**Comment by kevinhwang91** [comment]
I think I was wrong. Extensions or users shouldn't get the aggregated result. 

Maybe configure the priority in source with extensions will relieve the redraw.

**Comment by oblitum** [comment]
I'm using the suggested conf for continuous preselect (though I don't like the fuzzy behavior in rust :-/).

I have one question. With that conf, I noticed that <kbd>Enter</kbd> is being consumed for item select, but I don't want that, I just want `<c-y>` for that. Is this some default mapping? It's problematic for me, specially with continuous preselect, it often blocks <kbd>Enter</kbd> key.

**Comment by chemzqm** [comment]
@oblitum `<CR>` is vim's behavior or remapped by yourself, not controlled by coc.nvim.

**Comment by oblitum** [comment]
@chemzqm I never do such mappings so it should be some hardcoded aspect of neovim's preselect implementation. I will check it further. It really makes the feature not attractive anymore.

**Comment by oblitum** [comment]
It must be preselect specific indeed, because it doesn't happen if I select the second item, it only happens on preselect.

**Comment by kevinhwang91** [comment]
> It must be preselect specific indeed, because it doesn't happen if I select the second item, it only happens on preselect.

Either `noselect` disabled or `enablePreselect` enabled will auto select item which will enter the second state in popupmenu.

`:h popupmenu-completion`
```
The behavior of the <Enter> key depends on the state you are in:
first state:	  Use the text as it is and insert a line break.
second state:	  Insert the currently selected match.
third state:	  Use the text as it is and insert a line break.

In other words: If you used the cursor keys to select another entry in the
list of matches then the <Enter> key inserts that match.  If you typed
something else then <Enter> inserts a line break.
```

Workaround by remap it. `inoremap <silent><expr> <CR> pumvisible() ? "\<C-e>\<CR>" : "\<CR>"`

**Comment by oblitum** [comment]
@kevinhwang91 thx for explanation. Sigh, so it's just a vim builtin map, for me it's annoying, I'll try to apply some fix like that.

**Comment by oblitum** [comment]
But it's weird that it states

> If you used the cursor keys to select another entry in the
> list of matches then the \<Enter\> key inserts that match

But I do use the cursor keys as I have commented, and second state doesn't happen, it only happens when I actually don't use it, and preselect item is picked due to its _default_ selection. It's completely weird behavior that \<Enter\> selects if it's from preselection, but if I move selection down just one item, them \<Enter\> just inserts a line break.

**Comment by oblitum** [comment]
Ah, no no, my bad, by "cursor keys" the help actually means just arrow keys, and indeed, when I do use arrow keys, \<Enter\> will select it always, any entry, not solely the preselected one. The thing is that I never use arrow keys, and when you use \<c-n\> / \<c-p\> for item selection, which is what I do, then \<Enter\> produces a line break instead of selection.

So confusing....

**Comment by chemzqm** [comment]
>  the issue is that preselect is implemented as if "cursor keys" were used to do the preselection

Yes, same as VSCode and neovim only.
